x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java

```diff
@@ -1137,6 +1152,209 @@ protected Expression rule(Expression e) {
+
+                // the Literal will be moved to the right of the comparison, but only if data-compatible with what's there
+                return typesCompatible.apply(bcLiteral.dataType(), opLiteral.dataType()) == false;
+            }
+
+            Expression doSimplify() {
+                // force float point folding for FlP field
+                Literal bcl = operation.dataType().isRational()
+                    ? Literal.of(bcLiteral, ((Number) bcLiteral.value()).doubleValue())
+                    : bcLiteral;
+
+                if (operation instanceof BinaryComparisonInvertible == false) { // should be an assertion?
+                    throw new QlIllegalArgumentException("Unexpected [" + operation.symbol() + "] operation object");
+                }
+                Expression bcRightExpression = ((BinaryComparisonInvertible) operation).binaryComparisonInverse()
+                    .create(bcl.source(), bcl, opRight);
+                bcRightExpression = safeMaybeFold(bcRightExpression);
+                return bcRightExpression == null ? comparison : comparison.replaceChildren(List.of(opLeft, bcRightExpression));
+            }
+
+            static Expression simplify(BinaryComparison comparison, BiFunction<DataType, DataType, Boolean> typesCompatible) {
+                ArithmeticOperation operation = (ArithmeticOperation) comparison.left();
+                // Use symbol comp: SQL operations aren't available in this package (as dependencies)
+                String opSymbol = operation.symbol();
+                // Modulo can't be simplified.
+                if (MOD.symbol().equals(opSymbol)) {
+                    return comparison;
+                }
+                SimplifyOperation simplify = null;
+                if (isMulOrDiv(opSymbol)) {
+                    simplify = new SimplifyMulDiv(comparison);
+                } else if (ADD.symbol().equals(opSymbol) || SUB.symbol().equals(opSymbol)) {
+                    simplify = new SimplifyAddSub(comparison);
+                }
+
+                return (simplify == null || simplify.cannotSimplify(typesCompatible)) ? comparison : simplify.doSimplify();
+            }
+
+            static boolean isMulOrDiv(String opSymbol) {
+                return MUL.symbol().equals(opSymbol) || DIV.symbol().equals(opSymbol);
+            }
+        }
+
+        private static class SimplifyAddSub extends SimplifyOperation {
+
+            SimplifyAddSub(BinaryComparison comparison) {
+                super(comparison);
+            }
+
+            @Override
+            boolean cannotSimplify(BiFunction<DataType, DataType, Boolean> typesCompatible) {
+                // no ADD/SUB with floating fields
+                return operation.dataType().isRational() || super.cannotSimplify(typesCompatible);
+            }
+
+            @Override
+            Expression doSimplify() {
+                if (SUB.symbol().equals(operation.symbol()) && opRight instanceof Literal == false) {
```

Comment:
Check the enums directly, don't use their names.

Reply:
Also is this condition correct? the else (parent doSimplify) is picked if it is a sub but the opRight is a literal

Reply:
> Check the enums directly, don't use their names.

See above.

> Also is this condition correct? the else (parent doSimplify) is picked if it is a sub but the opRight is a literal

It is correct: in the else-case the literal has simply been "normalised" to the right and the simplification can be attempted.
The check you're referring to (in the child `SimplifyAddSub`) is there to avoid transforming the expression that ultimately still can't be simplified; ex.: for `1 - x > -MAX`, w/o the check, the next transformation would be `x - MAX < 1` and then it would stop there, since `MAX + 1` would overflow. That's what the comment on the first line below the last in the snippet above indicates.
But if the literal is on the right already (`x - 1 ...`), the optimisation can be attempted in the parent `doSimplify` (as per previous naming).